src/lib/homestore_backend/pg_blob_iterator.cpp

```diff
@@ -0,0 +1,128 @@
+    return iter->second.get();
+}
+
+void HSHomeObject::PGBlobIterator::create_pg_shard_snapshot_data(sisl::io_blob_safe& meta_blob) {
+    auto pg = get_pg_metadata();
+    auto& pg_info = pg->pg_info_;
+    auto& pg_shards = pg->shards_;
+
+    flatbuffers::FlatBufferBuilder builder;
+    std::vector< std::uint8_t > uuid(pg_info.replica_set_uuid.size());
+    std::copy(pg_info.replica_set_uuid.begin(), pg_info.replica_set_uuid.end(), uuid.begin());
+    auto pg_entry = CreatePGInfoEntry(builder, pg_info.id, 0 /* priority*/, builder.CreateVector(uuid));
+
+    std::vector< ::flatbuffers::Offset< ShardInfoEntry > > shard_entries;
+    for (auto& shard : pg_shards) {
+        auto& shard_info = shard->info;
+        // TODO add lsn.
+        shard_entries.push_back(CreateShardInfoEntry(
+            builder, static_cast< uint8_t >(shard_info.state), shard_info.placement_group, shard_info.id,
+            shard_info.total_capacity_bytes, shard_info.created_time, shard_info.last_modified_time));
+    }
+    builder.FinishSizePrefixed(CreateResyncPGShardInfo(builder, pg_entry, builder.CreateVector(shard_entries)));
+    meta_blob = sisl::io_blob_safe{builder.GetSize()};
+    std::memcpy(meta_blob.bytes(), builder.GetBufferPointer(), builder.GetSize());
+}
+
+int64_t HSHomeObject::PGBlobIterator::get_next_blobs(uint64_t max_num_blobs_in_batch, uint64_t max_batch_size_bytes,
+                                                     std::vector< BlobInfoData >& blob_data_vec, bool& end_of_shard) {
+    end_of_shard = false;
+    uint64_t total_bytes = 0, num_blobs = 0;
+    while (true) {
+        auto r = home_obj_.query_blobs_in_shard(pg_id_, cur_shard_seq_num_, cur_blob_id_ + 1, max_num_blobs_in_batch);
+        if (!r) { return -1; }
+        auto& blob_info_vec = r.value();
+        for (auto& info : blob_info_vec) {
+            if (info.pbas == HSHomeObject::tombstone_pbas) {
+                // Skip deleted blobs
+                continue;
+            }
+            auto result = home_obj_
+                              ._get_blob_data(repl_dev_, info.shard_id, info.blob_id, 0 /*start_offset*/,
+                                              0 /* req_len */, info.pbas)
+                              .get();
+            if (!result) {
+                LOGE("Failed to retrieve blob for shard={} blob={} pbas={}", info.shard_id, info.blob_id,
+                     info.pbas.to_string(), result.error());
+                return -1;
+            }
+
+            auto& blob = result.value();
+            num_blobs++;
+            total_bytes += blob.body.size() + blob.user_key.size();
+            if (num_blobs > max_num_blobs_in_batch || total_bytes > max_batch_size_bytes) { return 0; }
+
+            BlobInfoData blob_data{{info.shard_id, info.blob_id, std::move(info.pbas)}, std::move(blob)};
+            blob_data_vec.push_back(std::move(blob_data));
+            cur_blob_id_ = info.blob_id;
+        }
+
+        if (blob_info_vec.empty()) {
```

Comment:
this is not right,  we may possibly get `max_num_blobs_in_batch` tombstones. We should based on `query_blobs_in_shard ` return us `r.value().size() == 0`

Reply:
We are skipping the tombstones at line 72.  Tombstone has tombstone pbas which has size but its values are 0

Reply:
yes I see we skip all tombstone in L72, so tombstone will not be put into `blob_data_vec`.  Later on in this line we check `blob_info_vec.empty()`.

so in the case when all entries we got from `query_blobs_in_shard`( L67) are tombstone, the `blob_data_vec`  will be empty then we mark the shard as finished.  This is not right, there can still be live blobs after the array of tombstones

Reply:
We check the results of index. If its empty, there will be no more blobs in that shard. Changed the name to index_results_vec to avoid confusion.

Reply:
oh sorry it is my bad,  I mixed `blob_data_vec` vs `blob_info_vec`